src/apps/filters/kRing.c

```diff
@@ -55,33 +55,61 @@ void doCell(H3Index h, int k, int printDistances) {
 }
 
 int main(int argc, char* argv[]) {
-    // check command line args
-    if (argc != 2 && argc != 3) {
-        fprintf(stderr, "usage: %s [k] [printDistances]\n", argv[0]);
-        exit(1);
-    }
-
+    bool helpArg = false;
     int k = 0;
-    if (!sscanf(argv[1], "%d", &k)) error("k must be an integer");
+    bool printDistances = false;
+    H3Index origin = 0;
+    bool originPresent;
+    Arg args[] = {
+        {.names = {"-h", "--help"},
+         .isHelp = true,
+         .valuePresent = &helpArg,
+         .helpText = "Show this help message"},
+        {.names = {"-k", NULL},
+         .required = true,
+         .scanFormat = "%d",
+         .valueName = "k",
+         .value = &k,
+         .helpText = "Radius of hexagons"},
+        {.names = {"-d", "--print-distances"},
+         .valuePresent = &printDistances,
+         .helpText = "Print distance from origin after index"},
+        {.names = {"-o", "--origin"},
+         .scanFormat = "%" PRIx64,
+         .valueName = "origin",
+         .value = &origin,
+         .valuePresent = &originPresent,
+         .helpText =
+             "Origin, or not specified to read origins from standard in"}};
 
-    int printDistances = 0;
-    if (argc > 2) {
-        if (!sscanf(argv[2], "%d", &printDistances))
-            error("printDistances must be an integer");
+    char* errorMessage = NULL;
+    char* errorDetails = NULL;
+    if (parseArgs(argc, argv, 4, args, &errorMessage, &errorDetails) ||
+        helpArg) {
+        printHelp(helpArg ? stdout : stderr, argv[0],
```

Comment:
Should this "print help if help arg or on arg parse failure" behavior just be built into `parseArgs`?

Reply:
It would certainly deduplicate this code, which will need to be in all filter applications. The trade-off is this places another responsibility in `parseArgs`, possibly making `parseArgs` harder to test as a unit, and I would prefer to avoid calling `exit` since it requires knowing `parseArgs` may never return.

Reply:
Agreed that `parseArgs` shouldn't call `exit`, but I think the functionality of printing out help messages is pretty standard in arg parsing libs, and you could avoid both the `printHelp` call and (I think?) the memory management for `errorMessage` and `errorDetails` here if they were included in `parseArgs` - then this would simply read 
```
if (parseArgs(argc, argv, 4, args) || helpArg) {
  return helpArg ? 0 : 1;
}
```
which would be much simpler and easier to deal with in the filters. Given that we'll need this in every filter, I think the code reduction and reduced maintenance make this totally worthwhile.

Reply:
Good point, I refactored the code for printing help into `utility.c`

Reply:
This looks much cleaner IMO - thanks!